### DISCUSSION TITLE
In obj2ifc.py, why is the folliwing line needed?

### DISCUSSION THREAD

**Discussion by Martin15135215** [IfcOpenShell/IfcOpenShell#2852] [General]
https://github.com/IfcOpenShell/IfcOpenShell/blob/381570cd9e9652d792cbac4539bc9076d1a6c554/src/blenderbim/scripts/obj2ifc-meshlab.py#L142

So what purpose does self.history do? Nowhere else is this variable needed.

**Comment by Moult** [comment]
At a glance you're right it's useless. Feel free to delete and if its actually useless a PR would be appreciated:)

**Reply by Martin15135215** [reply]
Coud we instead of deleting https://github.com/IfcOpenShell/IfcOpenShell/blob/381570cd9e9652d792cbac4539bc9076d1a6c554/src/blenderbim/scripts/obj2ifc-meshlab.py#L142

delte the follwing line https://github.com/IfcOpenShell/IfcOpenShell/blob/381570cd9e9652d792cbac4539bc9076d1a6c554/src/blenderbim/scripts/obj2ifc-meshlab.py#L108-L109

and do for all the created IfcPorduct this `IfcBuildingElementProxy.OwnerHistory = self.history`. also the same for site, building etc.

because right now this script produce for each create_entity api instance a new IfcOwnerHistory -> with the new method, we would just reference one IfcOwnerHistory

i don't really see a difference between the many IfcOwnerHistory

![image](https://user-images.githubusercontent.com/110968398/224548728-1f05ae04-81d2-4f33-948d-83cf479fadea.png)

However there could be something that i am missing here...

**Reply by Moult** [reply]
Owner histories are applied to any rooted element, which also include relationships. Yes in theory you could bulk assign everything to a single history. However the API is designed to update the history as you continue to author.

I guess this incremental history updating has little value in a bulk script, so a better approach would be to have a way to 'turn off' history tracking, or make history tracking transaction based instead of per API call. That would be a better generic solution than something specific to the obj2ifc script. Would you like to have a shot at this? If not, creating a bug to track this would be very appreciated.

**Reply by Moult** [reply]
Note that there is also a performance penalty that I've noticed when a single history is shared by many elements (typically seen in exports from proprietary software). It makes going through inverses very, very slow.

**Reply by Martin15135215** [reply]
i heard from @berndhahnebach, that if in the api call, there is already the self.histrory variable as a parameter given, it does not create new IfcOwnerHistroy

**Reply by berndhahnebach** [reply]
> Note that there is also a performance penalty that I've noticed when a single history is shared by many elements (typically seen in exports from proprietary software). It makes going through inverses very, very slow.

I did not know about this performance issue. FreeCAD as well does use only one OwnerHistory for all products. Is it very slow for alle kinde of inverse operations? Would it makes sense to create a owner history foreach product on FreeCAD ifc export? cheers bernd

**Reply by Moult** [reply]
> i heard from @berndhahnebach, that if in the api call, there is already the self.histrory variable as a parameter given, it does not create new IfcOwnerHistroy

This is not correct unfortunately.

> Is it very slow for alle kinde of inverse operations?

It is slow whenever you need to get the inverse of the owner history. This used to be a problem, but maybe it has already been solved:

 - https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/owner/update_owner_history.py#L75
 - https://github.com/IfcOpenShell/IfcOpenShell/commit/ce3490824d3604f64a93762a246c57d245b7bb82 <-- more details about performance issue

Note that in that scenario we can get by with using `get_total_inverses` but maybe that might not work for other scenarios. As FreeCAD slowly gains more native IFC capabilities, I suspect the owner history will also evolve to be transaction based, like mentioned above.

**Comment by Martin15135215** [comment]
> Would you like to have a shot at this? If not, creating a bug to track this would be very appreciated.

I created the [issue](https://github.com/IfcOpenShell/IfcOpenShell/issues/2867) and i can try it at weekend, however i am not an python expert, still learning my ways
